sklearn/utils/extmath.py

```diff
@@ -1050,11 +1057,18 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
     result
         The output of the accumulator function passed to this function.
     """
-    if np.issubdtype(x.dtype, np.floating) and x.dtype.itemsize < 8:
-        result = op(x, *args, **kwargs, dtype=np.float64)
-    else:
-        result = op(x, *args, **kwargs)
-    return result
+    xp, _, x_device = get_namespace_and_device(x)
+    if xp.isdtype(x.dtype, "real floating"):
+        # Promote all dtypes, electing not to make copies if no dtype change is needed.
+        #  Normally we can rely on type promotion in `op`, but it isn't reliable for
+        #  float16 inputs.
+        target_dtype = max_precision_float_dtype(xp, device=x_device)
+        x = xp.astype(x, target_dtype, copy=False)
```

Comment:
Changing the implementation to look like:

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        kwargs["dtype"] = max_precision_float_dtype(xp, device=x_device)
    return op(x, *args, **kwargs)
```

causes test failures with torch due to `xp.matmul` not accepting a `dtype` parameter, and torch's default dtype being `float32`. `xp.matmul` does type promotion based on its arguments, but this is rather restrictive in the array api spec (e.g. no support for float16, can't promote int -> float).

This implementation fixes it for `torch.float{32,64}`, but not for `torch.float16` since the `array_api_compat` library doesn't step in to do type promotion with float16.

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        # We need to upcast. Some ops support this natively; others don't.
        target_dtype = max_precision_float_dtype(xp, device=x_device)

        def convert_(arr):
            return xp.astype(arr, target_dtype, copy=False)

        try:
            return op(x, *args, **kwargs, dtype=target_dtype)
        except TypeError:
            # This op doesn't support a dtype kwarg, it seems. Rely on manual
            # type promotion, at the cost of memory allocations.
            x = convert_(x)
    return op(x, *args, **kwargs)
```

Manually converting arguments fixes it for the `float16` usecase (or when using other array-api non-standard dtypes):

```python
def _safe_accumulator_op(op, x, *args, **kwargs):
    xp, _, x_device = get_namespace_and_device(x)
    if xp.isdtype(x.dtype, "real floating") and xp.finfo(x.dtype).bits < 64:
        # We need to upcast. Some ops support this natively; others don't.
        target_dtype = max_precision_float_dtype(xp, device=x_device)

        def convert_(arr):
            return xp.astype(arr, target_dtype, copy=False)

        try:
            return op(x, *args, **kwargs, dtype=target_dtype)
        except TypeError:
            # This op doesn't support a dtype kwarg, it seems. Rely on manual
            # type promotion, at the cost of memory allocations.
            x = convert_(x)
            args = [
                (convert_(arg) if hasattr(arg, "dtype") else arg) for arg in args
            ]
    return op(x, *args, **kwargs)
```

I'd go with the last implementation since it allows for the more efficient code path when you can pass a dtype to the operation (like numpy and cupy), while still supporting usecases where manual type promotion is required. Let me know what you think ~